### DISCUSSION TITLE
Which Oracle GraalVM 21 Preview features will be included?

### DISCUSSION THREAD

**Discussion by hrstoyanov** [oracle/graal#6988] [Q&A] [answered]
Hello,
Which of the Java 21 Preview features will be shipped with the corresponding Oracle GraalVM 21? I am particularly interested if GraalVM  21 will include :

  **442: |Foreign Function & Memory API (Third Preview)**

but also:

  444: | Virtual Threads
  446: | Scoped Values (Preview)
  448: | Vector API (Sixth Incubator)
  453: | Structured Concurrency (Preview)

**Comment by fniephaus** [comment, marked as answer]
Hi @hrstoyanov,

Thanks for starting this discussion. Here's what I found out after going through merged PRs and open tickets for the release of GraalVM for JDK 21:

- **Available for AOT and JIT**: Virtual Threads (444), Scope Values (446), and Structured Concurrency (453)
- **Work in progress**: Foreign Function & Memory API (442, *we are currently working on support for down calls first*)
- **Planned, but not yet implemented for JIT and AOT**: Vector API (448, *it might work, but not very fast yet*)

**Reply by plokhotnyuk** [reply]
How about Generational ZGC?

**Reply by fniephaus** [reply]
That is currently being worked on for JIT, no plans for AOT (yet).
